### ISSUE TITLE
greenlet is not installed on Apple M1

### ISSUE THREAD

**Issue by expobrain** [sqlalchemy/sqlalchemy#7714]
### Describe the bug

When installing SQLAlchemy 1.4.x with the `asyncio` extra on a Apple M1 machine the `greenlet` package is not installed even if it's listed in the requirements and for `aarch64` architectures.

### To Reproduce

```python
# On a Apple M1, inside a Docker container running Python in a linux distro
$ docker run --rm -it python:3.10 /bin/bash
root@afbf52f73731:/# pip install sqlalchemy==1.4.31
root@afbf52f73731:/# pip list | grep greenlet
greenlet   1.1.2

# On a Apple M1, in the local environment
$ pip install sqlalchemy==1.4.31
$ pip list | grep greenlet
# no output
```

### Error

```
.venv/lib/python3.10/site-packages/alembic/command.py:320: in upgrade
    script.run_env()
.venv/lib/python3.10/site-packages/alembic/script/base.py:563: in run_env
    util.load_python_file(self.dir, "env.py")
.venv/lib/python3.10/site-packages/alembic/util/pyfiles.py:92: in load_python_file
    module = load_module_py(module_id, path)
.venv/lib/python3.10/site-packages/alembic/util/pyfiles.py:108: in load_module_py
    spec.loader.exec_module(module)  # type: ignore
migrations/env.py:86: in <module>
    asyncio.run(run_migrations_online())
../../../.pyenv/versions/3.10.2/lib/python3.10/asyncio/runners.py:44: in run
    return loop.run_until_complete(main)
../../../.pyenv/versions/3.10.2/lib/python3.10/asyncio/base_events.py:641: in run_until_complete
    return future.result()
migrations/env.py:71: in run_migrations_online
    engine_from_config(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/create.py:755: in engine_from_config
    return create_engine(url, **options)
.venv/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py:309: in warned
    return fn(*args, **kwargs)
.venv/lib/python3.10/site-packages/sqlalchemy/engine/create.py:672: in create_engine
    event.listen(pool, "connect", on_connect)
.venv/lib/python3.10/site-packages/sqlalchemy/event/api.py:115: in listen
    _event_key(target, identifier, fn).listen(*args, **kw)
.venv/lib/python3.10/site-packages/sqlalchemy/event/registry.py:232: in listen
    self.dispatch_target.dispatch._listen(self, *args, **kw)
.venv/lib/python3.10/site-packages/sqlalchemy/pool/events.py:63: in _listen
    event_key.base_listen(**kw)
.venv/lib/python3.10/site-packages/sqlalchemy/event/registry.py:270: in base_listen
    for_modify._set_asyncio()
.venv/lib/python3.10/site-packages/sqlalchemy/event/attr.py:274: in _set_asyncio
    self._exec_once_mutex = AsyncAdaptedLock()
.venv/lib/python3.10/site-packages/sqlalchemy/util/concurrency.py:67: in AsyncAdaptedLock
    _not_implemented()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def _not_implemented():
        # this conditional is to prevent pylance from considering
        # greenlet_spawn() etc as "no return" and dimming out code below it
        if have_greenlet:
            return None
    
        if not compat.py3k:
            raise ValueError("Cannot use this function in py2.")
        else:
>           raise ValueError(
                "the greenlet library is required to use this function."
                " %s" % greenlet_error
                if greenlet_error
                else ""
            )
E           ValueError: the greenlet library is required to use this function. No module named 'greenlet'

.venv/lib/python3.10/site-packages/sqlalchemy/util/concurrency.py:47: ValueError
``` 

### Versions

- OS: OS X 12.2
- Python: 3.10
- SQLAlchemy: 1.4.31
- Database: not applicable
- DBAPI not applicable

### Additional context

The issue seems related to the way `platform.machine()` reports the underlying architecture, on Linux it reports `aarch64` and on OS X reports `arm64`, but the requirements contains only `aarch64` for `greenlet` https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_31/setup.cfg#L44

**Comment by CaselIT** [comment]
How nice of python to be consistent..

Thanks for reporting.

As a workaround you can install `sqlalchemy[asyncio]` that marks greenlet as required without conditions

**Comment by expobrain** [comment]
@CaselIT that's a nice solution, but works only with Pip and not on Poetry (and of course we use Poetry in our projects :-) )

**Comment by CaselIT** [comment]
strange that poetry does not let you specify an extra. from a quick search it seems it somewhat does https://python-poetry.org/docs/master/pyproject/#extras but I've never used it, so it may not actually work

**Comment by expobrain** [comment]
@CaselIT I mean, it does and this is the resulting entry in `pyproject.toml`:

```
SQLAlchemy = {extras = ["asyncio"], version = "^1.4.31"}
```

However in the `poetry.lock` it adds this line (which I reckon is coming from SQLAlchemy's `setup.cfg`):

```
greenlet = {version = "!=0.4.17", markers = "python_version >= \"3\" and (platform_machine == \"aarch64\" or platform_machine == \"ppc64le\" or platform_machine == \"x86_64\" or platform_machine == \"amd64\" or platform_machine == \"AMD64\" or platform_machine == \"win32\" or platform_machine == \"WIN32\")"}
```

which prevents `greenlet` to be installed.

**Comment by zzzeek** [comment]
there was a bug in poetry regarding this whole thing maybe they've fixed?  you can always list "greenlet" as an explicit dependency in your project

**Comment by CaselIT** [comment]
> @CaselIT I mean, it does and this is the resulting entry in `pyproject.toml`:
> 
> ```
> SQLAlchemy = {extras = ["asyncio"], version = "^1.4.31"}
> ```
> 
> However in the `poetry.lock` it adds this line (which I reckon is coming from SQLAlchemy's `setup.cfg`):
> 
> ```
> greenlet = {version = "!=0.4.17", markers = "python_version >= \"3\" and (platform_machine == \"aarch64\" or platform_machine == \"ppc64le\" or platform_machine == \"x86_64\" or platform_machine == \"amd64\" or platform_machine == \"AMD64\" or platform_machine == \"win32\" or platform_machine == \"WIN32\")"}
> ```
> 
> which prevents `greenlet` to be installed.

That line seems not correct, I guess poetry does not manages the case where a dependency is listed  in an extra with different markers.
Not sure if it's worth reporting.

From a library maintenance point of view poetry is a bit finichy, but I guess it may just be because it's not pip

**Comment by zzzeek** [comment]
how about this question.  when you "pip install greenlet" on the M1 arch, which file gets downloaded here?  https://pypi.org/project/greenlet/#files ?  it downloads the aarch64 files?

**Comment by zzzeek** [comment]
OK, there's an arm64 wheel format, such as seen here: https://pypi.org/project/numpy/#files  `numpy-1.22.2-cp310-cp310-macosx_11_0_arm64.whl `.  greenlet does not publish arm64 right now on pypi.  This is the criteria that we use to determine what goes in those requirements; that there's a wheel already available for greenlet.   So for the moment we would not add arm64 to this list until greenlet supplies a build.

**Comment by expobrain** [comment]
@zzzeek @CaselIT thank you for the updates.

What should I do with the PR, should I close it or leave it for when an `arm64` wheel will be available for `greenlet`?

**Comment by zzzeek** [comment]
do you have any appetite to appeal to greenlet to add an arm64/OSX wheel?  Im not sure where their wheel infrastructure comes from.   if they use github actions like we do...does GH actions support arm64/OSX ?

**Comment by CaselIT** [comment]
> does GH actions support arm64/OSX ?

last time I checked ~ 2 month ago no. Still no it seems https://github.com/actions/virtual-environments/issues/2187

**Comment by joloppo** [comment]
Using greenlet<2 fixed this for me.

**Comment by jenstroeger** [comment]
I think I came across a related scenario that puzzles me. On a x64 macOS machine I install SQLA into an empty venv:
```
> pip install sqlalchemy
...
Installing collected packages: typing-extensions, greenlet, sqlalchemy
Successfully installed greenlet-3.2.3 sqlalchemy-2.0.41 typing-extensions-4.14.0
```
and then the same on an arm64 macOS machine:
```
> pip install sqlalchemy
...
Installing collected packages: typing-extensions, sqlalchemy
Successfully installed sqlalchemy-2.0.41 typing-extensions-4.14.0
```
and I can’t figure out why the greenlet package is getting installed even though none I did _not_ specify any one of the dependent extras. I’ll keep digging but if you have a suggestion…

**Comment by zzzeek** [comment]
greenlet is still in the main requires for version 2.0.  it's in 2.1 that it's fully optional

**Comment by jenstroeger** [comment]
> greenlet is still in the main requires for version 2.0. it's in 2.1 that it's fully optional

Oh… the `main` branch is the current development branch 🤦🏻‍♂️ In `rel_2_0` and other release branches I see greenlet listed as a default dependency. I’m less puzzled now. Ok so the thread makes more sense now.

**Comment by paulocoutinhox** [comment]
I have the same error today: greenlet

`sqlalchemy>=2.0.43`

**Comment by paulocoutinhox** [comment]
I solved with this `sqlalchemy[asyncio]>=2.0.43`

**Comment by CaselIT** [comment]
That's the correct way of solving it, since 2.1 will no longer includ greenlet by default, and to use the asyncio part the corresponding extra needs to be installed
